keps/sig-api-machinery/20190603-immutable-fields.md

```diff
@@ -0,0 +1,202 @@
+---
+title: Immutable Fields
+authors:
+  - "@apelisse"
+owning-sig: sig-api-machinery
+participating-sigs:
+  - sig-api-machinery
+reviewers:
+approvers:
+editor: TBD
+creation-date: 2019-06-03
+last-updated: 2019-06-03
+status: provisional
+see-also:
+  - "/keps/sig-api-machinery/0006-apply.md"
+---
+
+# Immutable fields
+
+## Release Signoff Checklist
+
+- [ ] kubernetes/enhancements issue in release milestone, which links to KEP (this should be a link to the KEP location in kubernetes/enhancements, not the initial KEP PR)
+- [ ] KEP approvers have set the KEP status to `implementable`
+- [ ] Design details are appropriately documented
+- [ ] Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] Graduation criteria is in place
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+## Summary
+
+A lot of fields in APIs tend to be "immutable", they can't be changed after
+creation. This is true for example for many of the fields in pods. There is
+currently no way to declaratively document that fields are immutable, and one
+has to rely on either built-in validation for core types, or have to build a
+validating webhooks for CRDs.
+
+Providing a new `// +immutable` marker would make the API more descriptive to
+users and help API developers by verifying these assertions automatically.
+
+## Motivation
+
+Kubernetes tend to encourage "immutable" objects rather than "delete and
+recreate" on changes approach. Users have expressed the need for automatically
```

Comment:
Better, but I'm still confused by `"immutable" objects`. What does it mean? That currently object fields can be updated with the object staying the same?

Reply:
"There are resources in Kubernetes which have immutable fields by design, i.e. after creation of an object, those fields cannot be mutated anymore. E.g. a pod's specification is (mostly?) unchangeable once it is created. To change the pod, it must be deleted, recreated and rescheduled. Users want to implement the same kind of read-only semantics for CustomResources."

Reply:
Yeah, that looks great

Reply:
That is extremely clear, thanks. Only thing IMO could be changed: `s/the same kind of read-only semantics/the same kind of semantics/`. "read-only" makes perfectly sense in this paragraph, but it could be confusing when reading the doc as a whole because elsewhere there are references to OpenAPI readOnly semantics and it's explicitly stated that it's not what is needed.